bot/utils/redis_dict.py

```diff
@@ -0,0 +1,136 @@
+from __future__ import annotations
+
+import json
+from collections.abc import MutableMapping
+from enum import Enum
+from typing import Dict, List, Optional, Tuple, Union
+
+import redis as redis_py
+
+from bot import constants
+
+ValidRedisKey = Union[str, int, float]
+JSONSerializableType = Optional[Union[str, float, bool, Dict, List, Tuple, Enum]]
+
+
+class RedisDict(MutableMapping):
+    """
+    A dictionary interface for a Redis database.
+
+    Objects created by this class should mostly behave like a normal dictionary,
+    but will store all the data in our Redis database for persistence between restarts.
+
+    Redis is limited to simple types, so to allow you to store collections like lists
+    and dictionaries, we JSON deserialize every value. That means that it will not be possible
+    to store complex objects, only stuff like strings, numbers, and collections of strings and numbers.
+    """
+
+    _namespaces = []
+    _redis = redis_py.Redis(
+        host=constants.Bot.redis_host,
+        port=constants.Bot.redis_port,
+    )  # Can be overridden for testing
+
+    def __init__(self, namespace: Optional[str] = None) -> None:
+        """Initialize the RedisDict with the right namespace."""
+        super().__init__()
+        self._has_custom_namespace = namespace is not None
+
+        if self._has_custom_namespace:
+            self._set_namespace(namespace)
+        else:
+            self.namespace = "global"
+
+    def _set_namespace(self, namespace: str) -> None:
+        """Try to set the namespace, but do not permit collisions."""
+        while namespace in self._namespaces:
+            namespace = namespace + "_"
+
+        self._namespaces.append(namespace)
+        self._namespace = namespace
+
+    def __set_name__(self, owner: object, attribute_name: str) -> None:
+        """
+        Set the namespace to Class.attribute_name.
+
+        Called automatically when this class is constructed inside a class as an attribute, as long as
+        no custom namespace is provided to the constructor.
+        """
+        if not self._has_custom_namespace:
+            self._set_namespace(f"{owner.__name__}.{attribute_name}")
+
+    def __repr__(self) -> str:
+        """Return a beautiful representation of this object instance."""
+        return f"RedisDict(namespace={self._namespace!r})"
+
+    def __eq__(self, other: RedisDict) -> bool:
+        """Check equality between two RedisDicts."""
+        return self.items() == other.items() and self._namespace == other._namespace
+
+    def __ne__(self, other: RedisDict) -> bool:
+        """Check inequality between two RedisDicts."""
+        return self.items() != other.items() or self._namespace != other._namespace
+
+    def __setitem__(self, key: ValidRedisKey, value: JSONSerializableType):
```

Comment:
Good catch. I don't see any particular reason why we shouldn't. I think initially I didn't do it because I hadn't come up with the idea of making a JSONSerializableType annotation yet, but now that we've got that, there are obvious candidates for all of these to have return types. I will fix it.